File: dataset/human/java/file_9694.java
import java.util.Scanner;

class Node {
	int id, left, right;
	int parent = -1;
	int sibling = -1;	// ???????????°
	int degree;			// ????????°
	int depth;
	int height;

	public Node(int id) {
		this.id = id;
	}

	@Override
	public String toString() {

		String type;
		if(this.parent == -1) {
			type = "root";
		} else if (this.degree == 0) {
			type = "leaf";
		} else {
			type = "internal node";
		}

		String str =
				String.format(
						"node %d: parent = %d, sibling = %d, degree = %d, depth = %d, height = %d, %s",
						this.id, this.parent, this.sibling, this.degree, this.depth, this.height, type);
		return str;
	}
}

class Main {

	static int treeDepth = 0;

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int n = sc.nextInt();
		Node[] bTree = new Node[n];

		for (int i = 0; i < n; i++) {
			bTree[i] = new Node(i);
		}

		for (int i = 0; i < n; i++) {
			int now = sc.nextInt();
			int left = sc.nextInt();
			int right = sc.nextInt();

			bTree[now].left = left;
			bTree[now].right = right;

			int deg = 0;
			if (left != -1) {
				deg++;
				bTree[left].parent = now;
			}
			if (right != -1) {
				deg++;
				bTree[right].parent = now;
			}
			if (left != -1 && right != -1) {
				bTree[left].sibling = right;
				bTree[right].sibling = left;
			}

			bTree[now].degree = deg;

		}
		sc.close();

		int root = -1;
		for(Node node : bTree) {
			if (node.parent == -1) {
				root = node.id;
				break;
			}
		}

		bTree[root].depth = 0;
		setDepth(root, bTree);
		setHeight(root, bTree);

		for (Node node : bTree) {
			System.out.println(node);
		}

	}

	private static void setHeight(int id, Node[] bTree) {
		int h1 = 0;
		int h2 = 0;

		Node now = bTree[id];

		if (now.left != -1) {
			setHeight(now.left, bTree);
			h1 = bTree[now.left].height + 1;
		}
		if (now.right != -1) {
			setHeight(now.right, bTree);
			h2 = bTree[now.right].height + 1;
		}

		now.height = Math.max(h1, h2);
	}

	private static void setDepth(int id, Node[] bTree) {
		Node now = bTree[id];

		if (now.left != -1) {
			bTree[now.left].depth = now.depth + 1;
			setDepth(now.left, bTree);
		}
		if (now.right != -1) {
			bTree[now.right].depth = now.depth + 1;
			setDepth(now.right, bTree);
		}

		treeDepth = Math.max(treeDepth, now.depth);
	}
}

File: dataset/human/c/file_9694.c
#include <stdio.h>
#define MAX 100000
#define NIL -1

struct Node {
    int parent;
    int left;
    int right;
};

struct Node gNode[MAX];
int gDepth[MAX];

void print(int node){
  int i,left;
  printf("node %d: ",node);
  printf("parent = %d, ",gNode[node].parent);
  printf("depth = %d, ",gDepth[node]);
    
  if(gNode[node].parent == NIL){
    printf("root, ");
  }
  else if(gNode[node].left == NIL){
    printf("leaf, ");
  }
  else{
    printf("internal node, ");
  }
  printf("[");
    
    for(i=0,left=gNode[node].left;left!=NIL;i++,left=gNode[left].right){
      if(i) {
	printf(", ");
      }
        printf("%d",left);
    }
    printf("]\n");
}

void rec(int node ,int parent){
    gDepth[node]=parent;
    if(gNode[node].right!=NIL){
        rec(gNode[node].right,parent);
    }
    if(gNode[node].left!=NIL){
        rec(gNode[node].left,parent+1);
    }
}

int main(){
  int nodeNum,nodeIndex,node,left,root,n,i,j;
    scanf("%d",&n);
    for(i = 0;i < n;++i){
        gNode[i].parent = gNode[i].left = gNode[i].right = NIL;
    }
    for(i = 0;i < n;++i){
      scanf("%d%d",&nodeIndex,&nodeNum);
        for(j=0;j<nodeNum;++j){
	  scanf("%d",&node);
            if(j == 0){
                gNode[nodeIndex].left = node;
            }
            else{
                gNode[left].right = node;
            }
            left = node;
            gNode[node].parent = nodeIndex;
        }
    }    
    for(i = 0;i < n;++i){
        if(gNode[i].parent == NIL){
            root = i;
        }
    }
    rec(root,0);
    for(i = 0;i < n;++i){
        print(i);
    }
    return 0;
}
